lib/iris/coords.py

```diff
@@ -2233,31 +2239,78 @@ def _guess_bounds(self, bound_position=0.5):
                 "before guessing new ones."
             )
 
-        if getattr(self, "circular", False):
-            points = np.empty(self.shape[0] + 2)
-            points[1:-1] = self.points
-            direction = 1 if self.points[-1] > self.points[0] else -1
-            points[0] = self.points[-1] - (self.units.modulus * direction)
-            points[-1] = self.points[0] + (self.units.modulus * direction)
-            diffs = np.diff(points)
+        if monthly or yearly:
+            dates = self.units.num2date(self.points)
+            lower_bounds = []
+            upper_bounds = []
+            months_and_years = []
+            if monthly:
+                for date in dates:
+                    if date.month == 12:
+                        lyear = date.year
+                        uyear = date.year + 1
+                        lmonth = 12
+                        umonth = 1
+                    else:
+                        lyear = uyear = date.year
+                        lmonth = date.month
+                        umonth = date.month + 1
+                    date_pair = (date.year, date.month)
+                    if date_pair not in months_and_years:
+                        months_and_years.append(date_pair)
+                    else:
+                        raise ValueError(
+                            "Cannot guess monthly bounds for a coordinate with multiple "
+                            "points in a month."
+                        )
+                    lower_bounds.append(date.__class__(lyear, lmonth, 1, 0, 0))
+                    upper_bounds.append(date.__class__(uyear, umonth, 1, 0, 0))
+                bounds = self.units.date2num(np.array([lower_bounds, upper_bounds]).T)
+                contiguous = np.ma.allclose(bounds[1:, 0], bounds[:-1, 1])
+            elif yearly:
+                for date in dates:
+                    year = date.year
+                    if year not in months_and_years:
+                        months_and_years.append(year)
+                    else:
+                        raise ValueError(
+                            "Cannot guess yearly bounds for a coordinate with multiple "
+                            "points in a year."
+                        )
+                    lower_bounds.append(date.__class__(date.year, 1, 1, 0, 0))
+                    upper_bounds.append(date.__class__(date.year + 1, 1, 1, 0, 0))
+                bounds = self.units.date2num(np.array([lower_bounds, upper_bounds]).T)
+                contiguous = np.ma.allclose(bounds[1:, 0], bounds[:-1, 1])
```

Comment:
..and de-indent these ones
```suggestion
            bounds = self.units.date2num(np.array([lower_bounds, upper_bounds]).T)
            contiguous = np.ma.allclose(bounds[1:, 0], bounds[:-1, 1])
```

Reply:
[7ad2cdc](https://github.com/SciTools/iris/pull/6090/commits/7ad2cdc69d3dd63c3fb5391d85a2ef2a0713767c)